File: items.rb
require 'mechanize'
require 'nokogiri'
require 'open-uri'
require 'logger'
require 'json'

USER_AGENT = 'Mac Safari'
LOOKUP_SLEEP = 60
SKIP_SOLDOUT = true
HOST = 'http://www.supremenewyork.com'
ALL_URL = "#{HOST}/shop/all"

Dir.mkdir('html') unless File.exists?('html')
Dir.mkdir('items') unless File.exists?('items')

agent = Mechanize.new { |agent|
  agent.user_agent_alias = USER_AGENT
}

puts "+---------------------------------------+"
puts "| Supreme Bot: Item Fetcher         0.2 |"
puts "+---------------------------------------+"
puts "| LOOKUP_SLEEP: #{LOOKUP_SLEEP}"
puts "| SKIP_SOLDOUT: #{SKIP_SOLDOUT}"
puts "+---------------------------------------+"


while 420 != 710 do
  puts "Loading: #{ALL_URL}\n\n"

  items = []

  start =  Time.now.to_f
  agent.get(ALL_URL) do |page|
    doc = Nokogiri::HTML(page.body)
    doc.search('article a').each do |link|
      itemURL = "#{HOST}#{link.attributes['href']}"

      next if link.inner_html =~ /sold out/ && SKIP_SOLDOUT

      ## TODO: only get the item if its not sold out -- determine on index
      agent.get(itemURL) do |item|
        doc2 = Nokogiri::HTML(item.body)

        key1 = ''
        key2 = ''
        title = ''
        style = ''
        price = ''
        soldout = true
        sizes = []

        keys = link.attributes['href'].to_s.match(/([A-Z0-9]+)\/([A-Z0-9]+)$/i)

        key1 = keys[1]
        key2 = keys[2]

        doc2.search('#details h1').each do |res|
          title = res.children.first
        end

        doc2.search('.style').each do |res|
          style = res.children.first
        end

        doc2.search('.price span').each do |res|
          price = res.children.first
        end

        if item.body.match('sold out')
          soldout = true
        end

        puts "#{title} (style: #{style}, price: #{price})"

        doc2.search('#s option').each do |res|
            id = res.attributes['value']
            name = res.children.first
            sizes.push(name)
            # sizes.push({ id: id.to_s, name: name.to_s })
        end

        doc2.search('input[type="submit"]').each do |btn|
          if btn.attributes['value'].to_s == 'add to cart'
            soldout = false
          end
        end

        if 420 != 710 # !soldout && !SKIP_SOLDOUT
          items.push({
            key1: key1,
            key2: key2,
            url: itemURL,
            title: title.to_s,
            style: style.to_s,
            price: price.to_s,
            sizes: sizes,
            soldout: soldout
          })

        end
      end
    end
  end

  timestamp = Time.now.to_s.gsub(' ','_')
  ended =  Time.now.to_f
  diff = sprintf("%.4f", ended - start)

  puts "+---------------------------------------+"
  puts "took #{diff} seconds"

  ## Compare last with current ##

  ## open most recent item file
  latest = Dir.glob("items/*.json").max_by {|f| File.mtime(f)}

  if latest
    latest_items = File.read(latest)

    if latest_items == JSON.dump(items)
      puts "DUPLICATE ITEM ENTRY FOUND (#{latest})"
      puts "+---------------------------------------+"
      sleep(LOOKUP_SLEEP)
      next
    end
  end

  path = 'items/'
  filename = "items-#{timestamp}---#{diff}.json"

  File.open("#{path}#{filename}", 'w') do |file|
    file.write JSON.dump(items)
  end

  puts "Wrote #{ filename }, items: #{items.length} sleeping: Zzzzz #{LOOKUP_SLEEP}"
  puts "+---------------------------------------+"

  sleep(LOOKUP_SLEEP)
end

File: items.js
// deps
const fetch = require('node-fetch');
const createThrottle = require('async-throttle');
const cheerio = require('cheerio').load;

const Sequelize = require('sequelize');
var sqlite3 = require('sqlite3').verbose();
var db = new sqlite3.Database(':memory:');

const THROTTLE_COUNT = 3;
const HOST = `http://www.supremenewyork.com`;
const ALL_URL = `${HOST}/shop/all`;

var sequelize = new Sequelize('supreme.db', '', '', {
  host: 'localhost',
  dialect: 'sqlite',

  pool: {
    max: 5,
    min: 0,
    idle: 10000,
  },

  storage: 'supreme.db',
});

const Item = sequelize.define('item', {
  id: {
    type: Sequelize.INTEGER,
    field: 'id',
    primaryKey: true,
  },
  key1: {
    type: Sequelize.STRING,
    field: 'key1',
  },
  key2: {
    type: Sequelize.STRING,
    field: 'key2',
  },
  url: {
    type: Sequelize.STRING,
    field: 'url',
  },
  title: {
    type: Sequelize.STRING,
    field: 'title',
  },
  style: {
    type: Sequelize.STRING,
    field: 'style',
  },
  price: {
    type: Sequelize.STRING,
    field: 'price',
  },
  sizes: {
    type: Sequelize.STRING,
    field: 'sizes',
  },
  soldout: {
    type: Sequelize.INTEGER,
    field: 'soldout',
  },
  epoch: {
    type: Sequelize.INTEGER,
    field: 'epoch',
  },
});

const throttle = createThrottle(THROTTLE_COUNT);

fetch(ALL_URL).then(res => res.text()).then(data => {
  const $ = cheerio(data);
  const items = $('article a');

  console.log(`Found ${items.length} items`);

  Promise.all(
    items.each((i, item) => {
      throttle(async () => {
        const itemURL = `${HOST}/${item.attribs['href']}`;

        const res = await fetch(itemURL);
        const data = await res.text();
        const $ = cheerio(data);

        const title = $('#details h1').html();
        const style = $('.style').html();
        const price = $('.price span').html();
        const soldout = 0; // TODO

        const options = [];
        $('#s option').each((i, option) => {
          options.push(option.children[0].data);
        });

        let opt_string = options.join(',');

        const m = itemURL.match(/([A-Z0-9]+)\/([A-Z0-9]+)$/i);
        const key1 = m[1];
        const key2 = m[2];

        console.log(itemURL);
        console.log(key1);
        console.log(key2);
        console.log(title);
        console.log(style);
        console.log(price);
        console.log(opt_string);

        Item.create({
          key1,
          key2,
          title,
          url: itemURL,
          style: style,
          sizes: opt_string,
          price: price,
          soldout: soldout,
          epoch: 69,
        });

        console.log('--------------------------------');

        return $('title').text();
      });
    })
  ).then(titles => console.log('Titles:', titles));
});
